src/ucp/core/ucp_request.c

```diff
@@ -405,25 +406,46 @@ void ucp_request_dt_invalidate(ucp_request_t *req, ucs_status_t status)
 
     ucs_assert(status != UCS_OK);
     ucs_assert(ucp_ep_config(ep)->key.err_mode != UCP_ERR_HANDLING_MODE_NONE);
-    ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_USER_MEMH));
 
-    req->send.ep                = NULL;
     req->send.invalidate.worker = worker;
     req->status                 = status;
 
-    if (req->send.state.dt.dt.contig.memh == NULL) {
+    if (memh == NULL) {
         ucp_request_complete_send(req, status);
         return;
     }
 
     invalidate_map = ucp_request_get_invalidation_map(ep);
     ucp_trace_req(req, "mem invalidate buffer md_map 0x%" PRIx64 "/0x%" PRIx64,
-                  invalidate_map, req->send.state.dt.dt.contig.memh->md_map);
-    ucp_memh_invalidate(context, req->send.state.dt.dt.contig.memh,
-                        ucp_request_mem_invalidate_completion, req,
-                        invalidate_map);
-    ucp_memh_put(req->send.state.dt.dt.contig.memh);
+                  invalidate_map, memh->md_map);
+    ucp_memh_invalidate(context, memh, ucp_request_mem_invalidate_completion,
+                        req, invalidate_map);
+}
+
+void ucp_request_datatype_iter_invalidate(ucp_request_t *req,
+                                          ucs_status_t status)
+{
+    ucs_assert(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG);
+    ucp_request_memh_invalidate(req, req->send.state.dt_iter.type.contig.memh,
+                                status);
+}
+
+void ucp_request_dt_invalidate(ucp_request_t *req, ucs_status_t status)
+{
+    ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
```

Comment:
minor: better to use assertv and print dt

Reply:
```suggestion
    ucs_assertv(UCP_DT_IS_CONTIG(req->send.datatype),
                "wrong datatype=0x%"PRIx64, req->send.datatype);
```